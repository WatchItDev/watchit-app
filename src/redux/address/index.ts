import { createSlice, PayloadAction } from '@reduxjs/toolkit';

interface AddressState {
  address: string;
  profileId: string;
  showRainbow: boolean;
}

const initialState: AddressState = {
  address: '',
  profileId: '',
  showRainbow: false,
};

const addressSlice = createSlice({
  name: 'address',
  initialState,
  reducers: {
    storeAddress(state, action: PayloadAction<{ address: string; profileId: string }>) {
      state.address = action.payload.address;
      state.profileId = action.payload.profileId;
    },

    toggleRainbow(state) {
      state.showRainbow = !state.showRainbow;
    },

  },
});

export const { storeAddress, toggleRainbow } = addressSlice.actions;
export default addressSlice.reducer;
