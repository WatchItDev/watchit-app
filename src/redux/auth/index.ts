import {createSlice, PayloadAction} from "@reduxjs/toolkit";

export interface AuthReducerState {
  session: any;
  isSessionLoading: boolean;
  isLoginModalOpen: boolean;
  balance: number;
  currentStep: number;
  isUpdatingMetadata: boolean;
  email: string;
}

const initialState: AuthReducerState = {
  session: null,
  isSessionLoading: false,
  isLoginModalOpen: false,
  balance: 0,
  currentStep: 0,
  isUpdatingMetadata: false,
  email: "",
};

const authSlice = createSlice({
  name: "auth",
  initialState,
  reducers: {
    openLoginModal: (state) => {
      state.isLoginModalOpen = true;
    },
    closeLoginModal: (state) => {
      state.isLoginModalOpen = false;
    },
    setAuthLoading: (state, action: PayloadAction<Pick<AuthReducerState, "isSessionLoading">>) => {
      state.isSessionLoading = action.payload.isSessionLoading;
    },
    setBalance: (state, action: PayloadAction<Pick<AuthReducerState, "balance">>) => {
      state.balance = action.payload.balance;
    },
    setProfileCreationStep: (state, action: PayloadAction<{step: number}>) => {
      state.currentStep = action.payload.step;
    },
    resetCurrentStep: (state) => {
      state.currentStep = 0;
    },
    setSession: (state, action: PayloadAction<{session: any}>) => {
      state.session = action.payload.session;
    },
    setEmail: (state, action: PayloadAction<string>) => {
      state.email = action.payload;
    },
    updateProfileData: (
      state,
      action: PayloadAction<{
        name: string;
        bio: string;
        profileImage?: string;
        backgroundImage?: string;
      }>,
    ) => {
      if (state.session && state.session.profile) {
        state.session.profile.metadata = {
          ...state.session.profile.metadata,
          displayName: action.payload.name,
          bio: action.payload.bio,
          picture: {
            optimized: {
              uri: action.payload.profileImage ? action.payload.profileImage : undefined,
            },
          },
          coverPicture: {
            optimized: {
              uri: action.payload.backgroundImage ? action.payload.backgroundImage : undefined,
            },
          },
        };
      }
    },
    setIsUpdatingMetadata: (state, action: PayloadAction<boolean>) => {
      state.isUpdatingMetadata = action.payload;
    },
  },
});

export const {
  openLoginModal,
  closeLoginModal,
  setAuthLoading,
  setBalance,
  setProfileCreationStep,
  resetCurrentStep,
  setSession,
  updateProfileData,
  setIsUpdatingMetadata,
  setEmail,
} = authSlice.actions;

export default authSlice.reducer;
