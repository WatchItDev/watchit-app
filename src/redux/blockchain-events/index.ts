import { createSlice, PayloadAction } from '@reduxjs/toolkit'
import { Log } from 'viem'

interface BlockchainEventsState {
  events: Log[];
}

const initialState: BlockchainEventsState = {
  events: [],
}

const blockchainEventsSlice = createSlice({
  name: 'blockchainEvents',
  initialState,
  reducers: {
    addBlockchainEvent(state, action: PayloadAction<Log>) {
      state.events.push(action.payload)
    },
    setBlockchainEvents(state, action: PayloadAction<Log[]>) {
      state.events = action.payload
    },
    clearBlockchainEvents(state) {
      state.events = []
    },
  },
})

export const { addBlockchainEvent, setBlockchainEvents, clearBlockchainEvents } =
  blockchainEventsSlice.actions

export default blockchainEventsSlice.reducer
