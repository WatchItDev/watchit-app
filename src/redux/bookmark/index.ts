import { createSlice, PayloadAction } from '@reduxjs/toolkit';
import { Post } from '@src/graphql/generated/graphql.ts';

export interface BookmarkReducerState {
  bookmarkPublications: Post[];
  hiddenBookmarks: Post[];
}

const initialState: BookmarkReducerState = {
  bookmarkPublications: [],
  hiddenBookmarks: [],
};

const bookmarkSlice = createSlice({
  name: 'bookmark',
  initialState,
  reducers: {
    // Add a post to bookmarks and remove it from hidden if it is
    addBookmark: (state, action: PayloadAction<Post>) => {
      const isAlreadyBookmarked = state.bookmarkPublications.some(
        (publication) => publication.id === action.payload.id
      );

      if (!isAlreadyBookmarked) {
        state.bookmarkPublications.push(action.payload);
        // Remove from hiddenBookmarks if exists
        state.hiddenBookmarks = state.hiddenBookmarks.filter(
          (publication) => publication.id !== action.payload.id
        );
      }
    },

    // Move a post to hiddenBookmarks instead of deleting it completely
    removeBookmark: (state, action: PayloadAction<string>) => {
      const publicationToHide = state.bookmarkPublications.find(
        (publication) => publication.id === action.payload
      );

      if (publicationToHide) {
        state.hiddenBookmarks.push(publicationToHide);
        state.bookmarkPublications = state.bookmarkPublications.filter(
          (publication) => publication.id !== action.payload
        );
      }
    },

    // Remove all bookmarks
    clearBookmarks: (state) => {
      state.bookmarkPublications = [];
    },
  },
});

export const { addBookmark, removeBookmark, clearBookmarks } = bookmarkSlice.actions;

export default bookmarkSlice.reducer;
