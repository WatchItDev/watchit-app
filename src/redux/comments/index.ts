import { createSlice, PayloadAction } from '@reduxjs/toolkit';
import { AnyPublication } from '@lens-protocol/api-bindings';

type PendingComment = AnyPublication & { uri: string };

export type CommentsReducerState = {
  refetchTriggerByPublication: {
    [publicationId: string]: number;
  };
  hiddenComments: AnyPublication[];
  counterLikes: { [publicationId: string]: number };
  comments: { [publicationId: string]: AnyPublication[] };
  pendingComments: { [publicationId: string]: PendingComment[] };
};

const initialState: CommentsReducerState = {
  refetchTriggerByPublication: {},
  hiddenComments: [],
  counterLikes: {},
  comments: {},
  pendingComments: {},
};

const commentsSlice = createSlice({
  name: 'comments',
  initialState,
  reducers: {
    refetchCommentsByPublication: (state, action: PayloadAction<string>) => {
      const publicationId = action.payload;
      if (!state.refetchTriggerByPublication[publicationId]) {
        state.refetchTriggerByPublication[publicationId] = 1;
      } else {
        state.refetchTriggerByPublication[publicationId] += 1;
      }
    },
    hiddeComment: (state, action: PayloadAction<AnyPublication>) => {
      state.hiddenComments.push(action.payload);
    },
    setCounterLikes: (state, action: PayloadAction<{ publicationId: string; likes: number }>) => {
      state.counterLikes[action.payload.publicationId] = action.payload.likes;
    },
    incrementCounterLikes: (state, action: PayloadAction<string>) => {
      const publicationId = action.payload;
      if (state.counterLikes[publicationId] !== undefined) {
        state.counterLikes[publicationId] += 1;
      }
    },
    decrementCounterLikes: (state, action: PayloadAction<string>) => {
      const publicationId = action.payload;
      if (state.counterLikes[publicationId] !== undefined) {
        state.counterLikes[publicationId] -= 1;
      }
    },
    addPendingComment: (
      state,
      action: PayloadAction<{ publicationId: string; comment: PendingComment }>
    ) => {
      const { publicationId, comment } = action.payload;
      if (!state.pendingComments[publicationId]) {
        state.pendingComments[publicationId] = [];
      }
      // Prepend new comment to the beginning of the list
      state.pendingComments[publicationId] = [comment, ...state.pendingComments[publicationId]];
    },
    removePendingComment: (
      state,
      action: PayloadAction<{ publicationId: string; commentId: string }>
    ) => {
      const { publicationId, commentId } = action.payload;

      // Delete the comment from the pending list
      state.pendingComments[publicationId] = state.pendingComments[publicationId].filter(
        (comment) => comment.id !== commentId
      );
    },
  },
});

export const {
  hiddeComment,
  refetchCommentsByPublication,
  setCounterLikes,
  incrementCounterLikes,
  decrementCounterLikes,
  addPendingComment,
  removePendingComment,
} = commentsSlice.actions;

export default commentsSlice.reducer;
