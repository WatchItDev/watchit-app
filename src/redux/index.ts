export { minibarInitialState } from "@redux/minibar";
export { drawerInitialState } from "@redux/drawer";
export { authInitialState } from "@redux/auth";
export { bookmarkInitialState } from "@redux/bookmark";
export { commentsInitialState } from "@redux/comments";
export { notificationsInitialState } from "@redux/notifications";
export { followersInitialState } from "@redux/followers";
export { addressInitialState } from "@redux/address";
export { blockchainEventsInitialState } from "@redux/blockchain-events";
export { transactionInitialState } from "@redux/transactions";
