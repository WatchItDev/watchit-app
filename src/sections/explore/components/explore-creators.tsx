import CarouselCreators from '@src/components/carousel/variants/carousel-creators.tsx';
import { filterHiddenProfiles } from '@src/libs/profile.ts';
import { useGetRecentUsersQuery } from '@src/graphql/generated/hooks.tsx';
import { User } from '@src/graphql/generated/graphql.ts';
import { ExploreCreatorsSkeleton } from '@src/sections/explore/components/explore-creators.skeleton.tsx';
import { LoadingFade } from '@src/components/LoadingFade.tsx';

// ----------------------------------------------------------------------

export const ExploreCreators = () => {
  const { data, loading } = useGetRecentUsersQuery({ variables: { limit: 50 } })

  // FilteredCompletedProfiles is an array of objects, each object has a metadata property and inside exists a displayName en bio property; filter the profiles that not have a displayName and bio property
  const filtered = data?.getRecentUsers?.filter(
    (profile: User) => profile?.displayName && profile?.bio
  );

  // Clear ###HIDDEN### profiles
  const filteredProfiles = filterHiddenProfiles(filtered);

  return (
    <LoadingFade loading={loading} skeleton={<ExploreCreatorsSkeleton />} delayMs={500}>
      {!!filteredProfiles?.length && (
        <CarouselCreators
          profiles={filteredProfiles}
          title="Latest creators"
          minItemWidth={250}
          maxItemWidth={400}
        />
      )}
    </LoadingFade>
  );
}
