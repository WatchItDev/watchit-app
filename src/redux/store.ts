import { type Store } from 'redux';
import { configureStore } from '@reduxjs/toolkit';
import rootReducer from '@src/redux/reducer';

export const createStore = (initialState = {}): Store => {
  return configureStore({
    reducer: rootReducer,
    devTools: true,
    preloadedState: initialState,
    middleware: (getDefaultMiddleware) =>
      getDefaultMiddleware({
        serializableCheck: false,
      }),
  });
};

export type RootState = ReturnType<typeof store.getState>;

export const store = createStore();
