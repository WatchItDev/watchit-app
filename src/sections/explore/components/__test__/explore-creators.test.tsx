import { describe, expect, it, vi } from "vitest";
import { render, screen } from "@testing-library/react";
import { ExploreCreators } from "../explore-creators";
import { Provider } from "react-redux";
import { store } from "@src/redux/store";
import { MemoryRouter } from "react-router";

vi.mock("@src/workers/backgroundTaskWorker?worker", () => {
  return {
    default: class {
      postMessage() {}
      terminate() {}
      addEventListener() {}
      removeEventListener() {}
    },
  };
});

vi.mock("@src/hooks/components/use-item-per-slide", () => {
  return {
    useItemsPerSlide: () => ({
      itemsPerSlide: 2,
      parentRef: {
        current: {
          offsetWidth: 600,
        },
      },
    }),
  };
});

vi.mock("@lens-protocol/react-web", () => ({
  useBookmarks: () => ({
    data: [],
  }),
  useExploreProfiles: () => ({
    data: [
      {
        id: "profile1",
        metadata: {
          displayName: "Creator 1",
          bio: "A cool creator",
        },
      },
      {
        id: "profile2",
        metadata: {
          displayName: "Creator 2",
          bio: "Another cool creator",
        },
      },
      {
        id: "hiddenProfile",
        metadata: {
          displayName: "###HIDDEN### Hidden Creator",
          bio: "This creator is hidden",
        },
      },
    ],
    loading: false,
  }),
  ExploreProfilesOrderByType: {
    LatestCreated: "LatestCreated",
  },
  LimitType: {
    Fifty: "Fifty",
  },
}));

const renderWithProviders = () => {
  return render(
    <Provider store={store}>
      <MemoryRouter>
        <ExploreCreators />
      </MemoryRouter>
    </Provider>,
  );
};

describe("Testing in the ExploreCreators component", () => {
  it("should render creators and match snapshot", async () => {
    const { container } = renderWithProviders();

    await screen.findByText("Creator 1");
    await screen.findByText("Creator 2");
    expect(container).toMatchSnapshot();
  });
  it("should render the correct number of creators", async () => {
    renderWithProviders();
    await screen.findByText("Creator 1");
    await screen.findByText("Creator 2");
    const creators = screen.getAllByText(/Creator/);
    expect(creators.length).toBe(2);
  });
  it("should not render hidden creators", async () => {
    renderWithProviders();
    await screen.findByText("Creator 1");
    await screen.findByText("Creator 2");
    expect(screen.queryByText("Hidden Creator")).not.toBeInTheDocument();
  });

  it("should dispatch loading state to Redux store", async () => {
    const dispatchSpy = vi.spyOn(store, "dispatch");
    renderWithProviders();
    await screen.findByText("Creator 1");
    await screen.findByText("Creator 2");
    expect(dispatchSpy).toHaveBeenCalledWith(
      expect.objectContaining({
        type: "loading/setExploreLoading",
        payload: { key: "creators", isLoading: false },
      }),
    );
    expect(dispatchSpy).toHaveBeenCalledTimes(1);
  });
  it("should not render duplicate creators", async () => {
    renderWithProviders();
    await screen.findByText("Creator 1");
    await screen.findByText("Creator 2");
    const creators = screen.getAllByText(/Creator/);
    expect(creators.length).toBe(2);
    expect(creators[0]).not.toEqual(creators[1]);
  });
});
