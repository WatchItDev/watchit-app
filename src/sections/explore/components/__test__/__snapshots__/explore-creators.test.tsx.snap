// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Testing in the ExploreCreators component > should render creators and match snapshot 1`] = `
<div>
  <div
    class="MuiPaper-root MuiPaper-elevation MuiPaper-rounded MuiPaper-elevation1 MuiCard-root css-bhp9pd-MuiPaper-root-MuiCard-root"
  >
    <div
      class="MuiCardHeader-root css-1pkklwm-MuiCardHeader-root"
    >
      <div
        class="MuiCardHeader-content css-1qbkelo-MuiCardHeader-content"
      >
        <span
          class="MuiTypography-root MuiTypography-h5 MuiCardHeader-title css-1qvr50w-MuiTypography-root"
        >
          Latest creators
        </span>
      </div>
      <div
        class="MuiCardHeader-action css-sgoict-MuiCardHeader-action"
      >
        <button
          class="MuiButtonBase-root MuiIconButton-root MuiIconButton-sizeMedium css-78trlr-MuiButtonBase-root-MuiIconButton-root"
          data-testid="leftButton"
          tabindex="0"
          type="button"
        >
          <svg
            class="tabler-icon tabler-icon-chevron-left "
            fill="none"
            height="24"
            stroke="currentColor"
            stroke-linecap="round"
            stroke-linejoin="round"
            stroke-width="2"
            viewBox="0 0 24 24"
            width="24"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M15 6l-6 6l6 6"
            />
          </svg>
          <span
            class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
          />
        </button>
        <button
          class="MuiButtonBase-root MuiIconButton-root MuiIconButton-sizeMedium css-78trlr-MuiButtonBase-root-MuiIconButton-root"
          data-testid="rightButton"
          tabindex="0"
          type="button"
        >
          <svg
            class="tabler-icon tabler-icon-chevron-right "
            fill="none"
            height="24"
            stroke="currentColor"
            stroke-linecap="round"
            stroke-linejoin="round"
            stroke-width="2"
            viewBox="0 0 24 24"
            width="24"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M9 6l6 6l-6 6"
            />
          </svg>
          <span
            class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
          />
        </button>
      </div>
    </div>
    <div
      class="MuiCardContent-root css-fd8ij7-MuiCardContent-root"
    >
      <div
        class="MuiBox-root css-8lmyt0"
      >
        <div
          class="slick-slider slick-initialized"
        >
          <div
            class="slick-list"
          >
            <div
              class="slick-track"
              style="opacity: 1; transform: translate3d(0px, 0px, 0px);"
            >
              <div
                aria-hidden="false"
                class="slick-slide slick-active slick-current"
                data-index="0"
                style="outline: none; width: 0px;"
                tabindex="-1"
              >
                <div>
                  <div
                    class="MuiBox-root css-0"
                  >
                    <div
                      class="MuiBox-root css-k008qs"
                    >
                      <div
                        class="MuiBox-root css-2mh0az"
                      >
                        <div
                          class="MuiBox-root css-urjiuj"
                        >
                          <span
                            class="component-image MuiBox-root css-1w0a4u"
                          >
                            <span
                              class="component-image-wrapper lazy-load-image-background blur"
                              style="background-image: url(/assets/placeholder.svg); background-size: 100% 100%; color: transparent; display: inline-block;"
                            >
                              <img
                                alt="recentuser"
                                class="MuiBox-root css-3j6ntj"
                                src="https://picsum.photos/seed/cover3.jpg/1920/820"
                              />
                            </span>
                          </span>
                          <div
                            class="MuiPaper-root MuiPaper-elevation MuiPaper-rounded MuiPaper-elevation1 MuiCard-root css-vihmqx-MuiPaper-root-MuiCard-root"
                          >
                            <div
                              class="MuiAvatar-root MuiAvatar-rounded css-13ubyhz-MuiAvatar-root"
                            >
                              <img
                                alt="RECENTUSER"
                                class="MuiAvatar-img css-1pqm26d-MuiAvatar-img"
                                src="https://api.dicebear.com/9.x/bottts-neutral/svg?seed=profile3.jpg"
                              />
                            </div>
                            <div
                              class="MuiBox-root css-7gc5b8"
                            >
                              <div
                                class="MuiListItemText-root MuiListItemText-multiline css-konndc-MuiListItemText-root"
                              >
                                <span
                                  class="MuiTypography-root MuiTypography-body1 MuiTypography-noWrap MuiListItemText-primary css-f69c18-MuiTypography-root"
                                >
                                  <div
                                    class="MuiBox-root css-geek62"
                                  >
                                    <p
                                      class="MuiTypography-root MuiTypography-body1 css-nl8q6o-MuiTypography-root"
                                    >
                                      Recent User
                                    </p>
                                    <div
                                      class="MuiBox-root css-1wqjpco"
                                    />
                                  </div>
                                </span>
                                <span
                                  class="MuiTypography-root MuiTypography-body2 MuiTypography-noWrap MuiListItemText-secondary css-uhfo76-MuiTypography-root"
                                >
                                  Recent User Bio
                                </span>
                              </div>
                            </div>
                          </div>
                        </div>
                      </div>
                    </div>
                    <div
                      class="MuiBox-root css-k008qs"
                    />
                  </div>
                </div>
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;
